Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs

```diff
@@ -309,6 +337,50 @@ internal void OverrideCorrelationIdLookupHelper(CorrelationIdLookupHelper correl
         }
 
         /// <summary>
+        /// A request must be tracked as Active in order for telemetry to be recorded within OnEndRequest().
+        /// <see cref="System.Web.Handlers.TransferRequestHandler">
+        /// TransferRequestHandler can create a Child request to route extension-less requests to a controller.
+        /// (ex: site/home -> site/HomeController.cs)
+        /// We do not want duplicate telemetry logged for both the Parent and Child requests, so the activeRequests will be created OnBeginRequest.
+        /// When the child request OnEndRequest, the id will be removed from this dictionary and telemetry will not be logged for the parent.
+        /// </see>
+        /// </summary>
+        /// <remarks>
+        /// Unit test projects cannot create an [internal] IIS7WorkerRequest object.
+        /// Without this object, we cannot modify the Request.Headers without throwing a PlatformNotSupportedException.
+        /// Unit tests will have to initialize the RequestIdHeader.
+        /// The second IF will ensure the id is added to the activeRequests.
+        /// </remarks>
+        private void EnsureIsActiveRequestSafe(HttpContext context)
+        {
+            string requestId;
+            if (context.Request.Headers[RequestIdHeader] == null)
+            {
+                requestId = Guid.NewGuid().ToString();
+                context.Request.Headers[RequestIdHeader] = requestId;
+            }
+            else
+            {
+                requestId = context.Request.Headers[RequestIdHeader];
+            }
+            
+            if (!activeRequests.ContainsKey(requestId))
+            {
+                activeRequests.TryAdd(requestId, 0);
+            }
+        }
+
+        private void EnsureIsActiveRequest(HttpContext context)
+        {
+            if (context.Request.Headers[RequestIdHeader] == null)
+            {
+                string requestId = Guid.NewGuid().ToString();
+                context.Request.Headers[RequestIdHeader] = requestId;
+                activeRequests.TryAdd(requestId, 0);
```

Comment:
please protect the collection from memory leak. I'd suggest to check collection size from time to time and purge the entire collecting if it overgrown some huge limit. See [here](https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/da068a346377663e76a9b382ddf49e73cf96a1e6/Src/Common/CorrelationIdLookupHelper.cs#L21-L24) for example